### DISCUSSION TITLE
Replace "Run all" button with "Restart and run all" button

### DISCUSSION THREAD

**Discussion by benlindsay** [microsoft/vscode-jupyter#9217] [New Ideas]
In the Jupyter toolbar, there's a Run All button 

![image](https://user-images.githubusercontent.com/9361152/156601979-7013e753-b9bd-4917-bfd5-cf4a86fac4ce.png)

which looks a lot like the Restart and Run All button in Jupyter Lab

![image](https://user-images.githubusercontent.com/9361152/156602183-113442bb-f880-444a-8a77-c594c7de7ab1.png)

My initial expectation was that those buttons would be analogous, but in VSCode, it does not restart the kernel or clear outputs. Having that button fail to clear outputs is harmful because state from the previous run can be reused. For example, say you have a notebook with 2 cells:

```python
print(foo)
```

and

```python
foo = 42
```

If you run just the second cell, then click "Run All", your notebook will not fail even though it should. I can't think of any reason someone should want to rerun a notebook without also restarting the kernel. As for the clearing outputs bit, there is a case to be made for not clearing outputs, because it can be nice to see what was previously there. However, 2 points in favor of clearing outputs are 1) it would match behavior expectations from Jupyter Lab, and 2) when outputs are cleared prior to rerunning, it's much easier to tell where a failure occurred when scrolling through the notebook. Harder to tell when old outputs remain.

I hacked together a solution by adding a multicommand keybinding to keybindings.json, but that really shouldn't be necessary for a critical functionality. My keybinding looks like this:

```json
{
  "key": "alt+r",
  "command": "extension.multiCommand.execute",
  "args": {
    "sequence": [
      "notebook.clearAllCellsOutputs",
      "jupyter.notebookeditor.restartkernel",
      "notebook.execute"
    ]
  },
  "when": "notebookEditable && activeEditor == 'workbench.editor.notebook'"
}
```

Thanks for all of y'all's work on this project!

**Comment by DonJayamanne** [comment]
> I can't think of any reason someone should want to rerun a notebook without also restarting the kernel

This is a good suggestion, I agree this button in Jupyter Lab & Notebook restarts and runs all the cells.
Will bring this up in our team meeting. Thanks

**Comment by greazer** [comment]
Agreed. Made an issue to track this suggestion. See https://github.com/microsoft/vscode-jupyter/issues/9242
